File: Published/SourceCode/Python/DemoComparator.py

from BaseAPI import Comparable
import functools 
import operator

class Person(Comparable):
    def __init__(self, given, family, birth):
        self.givenName  = given
        self.familyName = family
        self.birthYear  = birth

    def __str__(self):
        return f"{self.givenName} {self.familyName} ({self.birthYear})"

    # ...as above...
    def __eq__(self, other): return self.familyName == other.familyName
    def __ne__(self, other): return self.familyName != other.familyName
    def __lt__(self, other): return self.familyName <  other.familyName
    def __le__(self, other): return self.familyName <= other.familyName
    def __gt__(self, other): return self.familyName >  other.familyName
    def __ge__(self, other): return self.familyName >= other.familyName


# Note: Python doesn't have comparators like Java does.
# The most similar is to define a comparator-like function:
def birthYearComparator(one, other):
    return (-1 if one.birthYear < other.birthYear else
             1 if one.birthYear > other.birthYear else 0)


def givenNameComparator(one, other):
    return (-1 if one.givenName < other.givenName else
             1 if one.givenName > other.givenName else 0)

def fullNameComparator(one, other):
    return (-1 if one.familyName < other.familyName else
             1 if one.familyName > other.familyName else
            -1 if one.givenName  < other.givenName  else
             1 if one.givenName  > other.givenName  else 0)

def getPeople():
    return [Person("Unsuk", "Chin", 1961),
            Person("Anna", "Thorvaldsdóttir", 1977),
            Person("Andrea", "Tarrodi", 1981),
            Person("Diana", "Čemerytė", 1974),
            Person("Elfrida", "Andrée", 1841),
            Person("Guy", "d’Hardelot", 1858),
            Person("Nadia", "Boulanger", 1887),
            Person("Lili", "Boulanger", 1893),
            ]

print("\n### No order");
people = getPeople()
for p in people: print(p)

print("\n### Natural ordering")
people = getPeople()  # reset the people list
people.sort()
for p in people: print(p)

print("\n### Ordered by birth year (pre-Java-8 solution)")
byBirthYear = functools.cmp_to_key(birthYearComparator)
people = getPeople()  # reset the people list
people.sort(key=byBirthYear)
for p in people: print(p)

print("\n### Ordered by birth year (functional solution)")
byBirthYear = lambda person: person.birthYear
people = getPeople()  # reset the people list
people.sort(key=byBirthYear)
for p in people: print(p)

print("\n### Ordered by birth year (using a key extractor)")
byBirthYear = operator.attrgetter('birthYear')
people = getPeople()  # reset the people list
people.sort(key=byBirthYear)
for p in people: print(p)

print("\n### Ordered by given name (pre-Java-8 solution)")
byGivenName = functools.cmp_to_key(givenNameComparator)
people = getPeople()  # reset the people list
people.sort(key=byGivenName)
for p in people: print(p)

print("\n### Ordered by given name (functional solution)")
byGivenName = lambda person: person.givenName
people = getPeople()  # reset the people list
people.sort(key=byGivenName)
for p in people: print(p)

print("\n### Ordered by given name (using a key extractor)")
byGivenName = operator.attrgetter('givenName')
people = getPeople()  # reset the people list
people.sort(key=byGivenName)
for p in people: print(p)

print("\n### Ordered by full name: family name + given name (pre-Java-8 solution)")
byFullName = functools.cmp_to_key(fullNameComparator)
people = getPeople()  # reset the people list
people.sort(key=byFullName)
for p in people: print(p)

print("\n### Ordered by full name: family name + given name (functional solution and tuples)")
# In Python we can simply create a tuple, and it will sort the way we want
byFullName = lambda person: (person.familyName, person.givenName)
people = getPeople()  # reset the people list
people.sort(key=byFullName)
for p in people: print(p)

print("\n### Ordered by Swedish locale, case-insensitive")
print("# Note: There's a bug in Python's Swedish locale, so Č comes after all other letters")
import locale
locale.setlocale(locale.LC_COLLATE, 'sv_SE')
bySwedishLocale = lambda person: (locale.strxfrm(person.familyName.casefold()),
                                      locale.strxfrm(person.givenName.casefold()))
# Note: There's a bug in Python's Swedish locale, so Č comes after all other letters
people = getPeople()  # reset the people list
people.sort(key=bySwedishLocale)
for p in people: print(p)
# Note: Because of a bug in Python's Swedish locale, Diana Čemerytė is still printed last

print("\n### Ordered by Swedish locale, given name first")
bySwedishLocale = lambda person: (locale.strxfrm(person.givenName.casefold()),
                                  locale.strxfrm(person.familyName.casefold()))
people = getPeople()  # reset the people list
people.sort(key=bySwedishLocale)
for p in people: print(p)

File: Published/SourceCode/Java/DemoComparator.java

import java.util.*;
import java.text.Collator;

class Person implements Comparable<Person> {
    private String givenName;
    private String familyName;
    private int birthYear;

    // Constructor
    Person(String given, String family, int birth) {
        givenName  = given;
        familyName = family;
        birthYear  = birth;
    }

    // Getters
    public String givenName()  { return givenName  ; }
    public String familyName() { return familyName ; }
    public int    birthYear()  { return birthYear  ; }

    // Pretty-printing
    public String toString() {
        return givenName() + " " + familyName() + " (" + birthYear() + ")";
    }

    // Natural ordering
    // ...as above...
    public int compareTo(Person other) {
        return this.familyName().compareTo(other.familyName());
    }
}


class BirthYearComparator implements Comparator<Person> {
    public int compare(Person one, Person other) {
        return Integer.compare(one.birthYear(), other.birthYear());
    }
}


class GivenNameComparator implements Comparator<Person> {
    public int compare(Person one, Person other) {
        return one.givenName().compareTo(other.givenName());
    }
}

class FullNameComparator implements Comparator<Person> {
    public int compare(Person one, Person other) {
        int cmp = one.familyName().compareTo(other.familyName());
        if (cmp != 0) return cmp;
        cmp = one.givenName().compareTo(other.givenName());
        if (cmp != 0) return cmp;
        return 0;
    }
}

public class DemoComparator {
static ArrayList<Person> getPeople() {
    ArrayList<Person> people = new ArrayList<>();
    people.add(new Person("Unsuk", "Chin", 1961));
    people.add(new Person("Anna", "Thorvaldsdóttir", 1977));
    people.add(new Person("Andrea", "Tarrodi", 1981));
    people.add(new Person("Diana", "Čemerytė", 1974));
    people.add(new Person("Elfrida", "Andrée", 1841));
    people.add(new Person("Guy", "d’Hardelot", 1858));
    people.add(new Person("Nadia", "Boulanger", 1887));
    people.add(new Person("Lili", "Boulanger", 1893));
    return people;
}
    
public static void main(String[] args) {
System.out.println("\n### No order");
ArrayList<Person> people = getPeople();
for (Person p : people) System.out.println(p);

System.out.println("\n### Natural ordering");
people = getPeople(); // reset the people list
Collections.sort(people);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by birth year (pre-Java-8 solution)");
Comparator<Person> byBirthYear = new BirthYearComparator();
people = getPeople(); // reset the people list
Collections.sort(people, byBirthYear);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by birth year (functional solution)");
byBirthYear = (one, other) -> Integer.compare(one.birthYear(), other.birthYear());
people = getPeople(); // reset the people list
Collections.sort(people, byBirthYear);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by birth year (using a key extractor)");
byBirthYear = Comparator.comparingInt(Person::birthYear);
people = getPeople(); // reset the people list
Collections.sort(people, byBirthYear);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by given name (pre-Java-8 solution)");
Comparator<Person> byGivenName = new GivenNameComparator();
people = getPeople(); // reset the people list
Collections.sort(people, byGivenName);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by given name (functional solution)");
byGivenName = (one, other) -> one.givenName().compareTo(other.givenName());
people = getPeople(); // reset the people list
Collections.sort(people, byGivenName);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by given name (using a key extractor)");
byGivenName = Comparator.comparing(Person::givenName);
people = getPeople(); // reset the people list
Collections.sort(people, byGivenName);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by full name: family name + given name (pre-Java-8 solution)");
Comparator<Person> byFullName = new FullNameComparator();
people = getPeople(); // reset the people list
Collections.sort(people, byFullName);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by full name: family name + given name (using key extractors and thenComparing)");
byFullName = Comparator.comparing(Person::familyName)
    .thenComparing(Person::givenName);
people = getPeople(); // reset the people list
Collections.sort(people, byFullName);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by Swedish locale, case-insensitive");
Collator swedishLocale = Collator.getInstance(new Locale("sv", "SE"));
swedishLocale.setStrength(Collator.PRIMARY);
Comparator<Person> bySwedishLocale = 
    Comparator.comparing(Person::familyName, swedishLocale)
    .thenComparing(Person::givenName, swedishLocale);
people = getPeople(); // reset the people list
Collections.sort(people, bySwedishLocale);
for (Person p : people) System.out.println(p);

System.out.println("\n### Ordered by Swedish locale, given name first");
Comparator<Person> bySwedishGivenName = 
    Comparator.comparing(Person::givenName, swedishLocale)
    .thenComparing(Person::familyName, swedishLocale);
people = getPeople(); // reset the people list
Collections.sort(people, bySwedishGivenName);
for (Person p : people) System.out.println(p);
}
}
